tools/tx-scanner/cmd/scan.go

```diff
@@ -0,0 +1,153 @@
+package cmd
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/sei-protocol/sei-chain/tools/tx-scanner/client"
+	"github.com/sei-protocol/sei-chain/tools/tx-scanner/query"
+	"github.com/sei-protocol/sei-chain/tools/tx-scanner/state"
+	"github.com/spf13/cobra"
+	"golang.org/x/time/rate"
+)
+
+func ScanCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "scan-tx",
+		Short: "A tool to scan missing transactions",
+		Run:   execute,
+	}
+	cmd.PersistentFlags().String("endpoint", "127.0.0.1", "GRPC server endpoint")
+	cmd.PersistentFlags().Int("port", 9090, "GRPC server port")
+	cmd.PersistentFlags().Int("batch-size", 100, "Batch size to query")
+	cmd.PersistentFlags().Int("bps-limit", 400, "Blocks per second limit")
+	cmd.PersistentFlags().Int64("start-height", 0, "Start height")
+	cmd.PersistentFlags().String("state-dir", "", "State file directory, the scanner will record the last scanned offset and scan results")
+	return cmd
+}
+
+func execute(cmd *cobra.Command, _ []string) {
+	endpoint, _ := cmd.Flags().GetString("endpoint")
+	port, _ := cmd.Flags().GetInt("port")
+	bpsLimit, _ := cmd.Flags().GetInt("bps-limit")
+	batchSize, _ := cmd.Flags().GetInt("batch-size")
+	stateDir, _ := cmd.Flags().GetString("state-dir")
+	startHeight, _ := cmd.Flags().GetInt64("start-height")
+	var badBlocks []int64
+	var currentState = state.State{}
+	if batchSize > bpsLimit {
+		batchSize = bpsLimit
+	}
+	if startHeight <= 0 {
+		startHeight = 1
+	}
+	if stateDir != "" {
+		scanState, _ := state.ReadState(stateDir)
+		if scanState.LastProcessedHeight > 0 {
+			fmt.Printf("Detected last processed height: %d\n", scanState.LastProcessedHeight)
+			currentState = scanState
+			startHeight = currentState.LastProcessedHeight
+			badBlocks = currentState.BlocksMissingTxs
+		}
+	}
+	fmt.Printf("Starting the scan from height: %d\n", startHeight)
+	client.InitializeGRPCClient(endpoint, port)
+	rateLimiter := rate.NewLimiter(rate.Limit(bpsLimit), bpsLimit)
+	latestHeight := getLatestBlockHeight()
+	var currBlockHeight = startHeight
+	for {
+		if currBlockHeight >= latestHeight {
+			time.Sleep(10 * time.Second)
+			latestHeight = getLatestBlockHeight()
+			continue
+		}
+		if !rateLimiter.AllowN(time.Now(), batchSize) {
+			time.Sleep(1 * time.Millisecond)
```

Comment:
nit - would it make sense to bump this up a bit more? like 10ms

Reply:
Sure thing